src/Components/Shared/src/BaseJSDataStream.cs

```diff
@@ -0,0 +1,198 @@
+                    throw new EndOfStreamException($"The incoming data chunk cannot be empty.");
+                }
+
+                _bytesRead += chunk.Length;
+
+                if (_bytesRead > _totalLength)
+                {
+                    throw new EndOfStreamException($"The incoming data stream declared a length {_totalLength}, but {_bytesRead} bytes were sent.");
+                }
+
+                await _pipe.Writer.WriteAsync(chunk, _streamCancellationToken);
+
+                if (_bytesRead == _totalLength)
+                {
+                    await CompletePipeAndDisposeStream();
+                }
+
+                return true;
+            }
+            catch (Exception e)
+            {
+                await CompletePipeAndDisposeStream(e);
+
+                // Fatal exception, crush the circuit. A well behaved client
+                // should not result in this type of exception.
+                if (e is EndOfStreamException)
+                {
+                    throw;
+                }
+
+                return false;
+            }
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => _totalLength;
+
+        public override long Position
+        {
+            get => _pipeReaderStream.Position;
+            set => throw new NotSupportedException();
+        }
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException("Synchronous reads are not supported.");
+
+        public override long Seek(long offset, SeekOrigin origin)
+            => throw new NotSupportedException();
+
+        public override void SetLength(long value)
+            => throw new NotSupportedException();
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var linkedCancellationToken = GetLinkedCancellationToken(_streamCancellationToken, cancellationToken);
+            return await _pipeReaderStream.ReadAsync(buffer.AsMemory(offset, count), linkedCancellationToken);
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            var linkedCancellationToken = GetLinkedCancellationToken(_streamCancellationToken, cancellationToken);
+            return await _pipeReaderStream.ReadAsync(buffer, linkedCancellationToken);
+        }
+
+        private static CancellationToken GetLinkedCancellationToken(CancellationToken a, CancellationToken b)
+        {
+            if (a.CanBeCanceled && b.CanBeCanceled)
```

Comment:
`CreateLinkedTokenSource` already does this. You don't need to do this here.

Reply:
@BrennanConroy raised a concern in a prior PR about how `LinkedCancellationToken` always allocates. Reviewing [the source](https://github.com/dotnet/runtime/blob/fe9a54df8f4b5e6a2e4cafc0e4234a64e8c9cf8e/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L740-L786), that does appear to be the case.